### PR TITLE
[extractor/tiktok] Improve `TikTokLive` extractor

### LINKED ISSUE #6459: [tiktok.com] Failure to obtain m3u8 URL of some live streams
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that a **supported** site is broken
- [X] I've verified that I'm running yt-dlp version **2023.03.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Ireland

### Provide a description that is worded well enough to be understood

There is an error when downloading some tiktok live streams (e.g. https://www.tiktok.com/@mikey_____7/live or  https://www.tiktok.com/@knkjewelrydesign/live)

> ERROR: [tiktok:live] 7207751234156120874: Failed to obtain m3u8 URL;

This error occurs not on every stream, but very often. And it is consistent with particular streams - after retrying yt-dlp fails to download m3u8 URL. For other streams, successful download is also consistent.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', 'https://www.tiktok.com/@mikey_____7/live']
[debug] Encodings: locale cp1251, fs utf-8, pref cp1251, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.03.04 [392389b7d] (win_exe)
[debug] Python 3.8.10 (CPython AMD64 64bit) - Windows-10-10.0.19045-SP0 (OpenSSL 1.1.1k  25 Mar 2021)
[debug] exe versions: ffmpeg 5.1.2-essentials_build-www.gyan.dev (setts), ffprobe 5.1.2-essentials_build-www.gyan.dev
[debug] Optional libraries: Cryptodome-3.17, brotli-1.0.9, certifi-2022.12.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-10.4
[debug] Proxy map: {}
[debug] Loaded 1786 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Available version: stable@2023.03.04, Current version: stable@2023.03.04
Current Build Hash: 5590c57bd0433ed239a2deaaf92e2ad6f37fe50f53664c821575cafe106a9421
yt-dlp is up to date (stable@2023.03.04)
[tiktok:live] Extracting URL: https://www.tiktok.com/@mikey_____7/live
[tiktok:live] mikey_____7: Downloading webpage
[tiktok:live] 7207751234156120874: Downloading JSON metadata
ERROR: [tiktok:live] 7207751234156120874: Failed to obtain m3u8 URL; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "yt_dlp\extractor\common.py", line 694, in extract
  File "yt_dlp\extractor\tiktok.py", line 1020, in _real_extract
  File "yt_dlp\extractor\common.py", line 1919, in _extract_m3u8_formats
  File "yt_dlp\extractor\common.py", line 1937, in _extract_m3u8_formats_and_subtitles
```

### PR DESCRIPTION
**Pull request by bashonly**
Credit to @mokrueger for sharing the necessary API endpoint. This new endpoint seems to provide stream URLs more reliably than the old endpoint, which this PR only uses as a fallback.

The PR also adds support for livestream mobile share links (`m.tiktok.com/share/live/{room_id}/`) and for streaming http-flv formats.

In my testing, as far as quality is concerned, it seems that `FULL_HD1 == rtmp == hls`. I've given HLS priority since most users probably prefer mp4 over flv.

I haven't found any examples where a stream URL is explicitly labeled as `ORIGION` (original quality), but I suspect that this format ID actually refers to the `rtmp` URL based on this pattern:
 * `FULL_HD1` URLs end with `-uhd.flv`
 * `HD1` URLs end with `-hd.flv`
 * `rtmp` URLs end with `-or4.flv`
 
Closes #6459

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/tiktok.py [hunk 1]
```diff
@@ -984,39 +984,119 @@ def _real_extract(self, url):
 
 
 class TikTokLiveIE(InfoExtractor):
-    _VALID_URL = r'https?://(?:www\.)?tiktok\.com/@(?P<id>[\w\.-]+)/live'
+    _VALID_URL = r'''(?x)
+                         https?://(?:
+                             (?:www\.)?tiktok\.com/@(?P<uploader>[\w.-]+)/live|
+                             m\.tiktok\.com/share/live/(?P<id>\d+)
+                         )'''
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:991]:
```suggestion
    _VALID_URL = r'''(?x)https?://(?:
        (?:www\.)?tiktok\.com/@(?P<uploader>[\w.-]+)/live|
        m\.tiktok\.com/share/live/(?P<id>\d+)
    )'''
```

File: yt_dlp/extractor/tiktok.py [hunk 2]
```diff
@@ -984,39 +984,119 @@ def _real_extract(self, url):
...
 
-    def _real_extract(self, url):
-        uploader = self._match_id(url)
-        webpage = self._download_webpage(url, uploader, headers={'User-Agent': 'User-Agent:Mozilla/5.0'})
-        room_id = self._html_search_regex(r'snssdk\d*://live\?room_id=(\d+)', webpage, 'room ID', default=None)
-        if not room_id:
-            raise UserNotLive(video_id=uploader)
-        live_info = traverse_obj(self._download_json(
-            'https://www.tiktok.com/api/live/detail/', room_id, query={
+    QUALITIES = ('SD1', 'SD2', 'HD1', 'pm_mt_video_720p60', 'FULL_HD1', 'pm_mt_video_1080p60', 'rtmp', 'ORIGION', 'hls')
```
Comment (pukkandan) [inline yt_dlp/extractor/tiktok.py:1025]:
```suggestion
    QUALITIES = qualities(('SD1', 'SD2', 'HD1', 'pm_mt_video_720p60', 'FULL_HD1', 'pm_mt_video_1080p60', 'rtmp', 'ORIGION', 'hls'))
```
since u never access the tuple directly

**Review by pukkandan — APPROVED** [review]

**Comment by bashonly** [comment]
@pukkandan I made some additional changes, maybe you want to review before merge

I decided it was a bad idea to skip webpage extraction if `room_id` can be extracted from the URL since the ID changes with each livestream. The mobile share links redirect to the user's live page where a fresh `room_id` can be extracted.

Some yt-dlp users have difficulty downloading TT webpages, so I made all webpage-related extraction non-fatal unless we can't get the `room_id`

**Review by pukkandan — APPROVED** [review]

**Comment by mokrueger** [comment]
@bashonly @pukkandan I wanna note that I've found a scenario where this endpoint is not providing a url. I think it will not break anything but I still wanted to share the info. If a user is a guest in a live then it will not provide anything since it only works on that user's specific webcast room. The previous endpoint (sometimes) provides the m3u8 file of the owner of the room where the user is a guest in. Just a random fyi.

Thanks for the PR btw ;)

**Comment by mokrueger** [comment]
@bashonly quick question. How often is the _extract_m3u8_formats endpoint called? Idk if its relevant, but this method downloads each m3u8 right and looks for additional information, I think(?)
In what I use privately I made another solution that uses the info from:
`alt_live_info["stream_url"]["live_core_sdk_data"]["pull_data"]["stream_data"]`
and for every entry in that thing, it does something along the lines of this (sorry for the mess but I think you can figure out what I mean):
```
sdk_params = json.loads(main.get("sdk_params"))
            common = {
                "vcodec": sdk_params.get("VCodec"),
                "resolution": sdk_params.get("resolution").lower(),
                "vbr": sdk_params.get("vbitrate"),
                "has_drm": None,
                "quality": None,
            }
            if main.get("hls"):
                f.append({
                    "format_id": f"hls-{sdk_params.get('resolution').lower()}",
                    "url": main.get("hls"),
                    "protocol": "m3u8_native",
                    "ext": "mp4",
                    **common,
                })
        .....
```
Long story short. I don't know if it is necessary to call the _extract_m3u8_formats, because I didn't see that it adds any more information and increases the amount of api calls (or website downloads) unnecessarily. But feel free to correct me :D

**Comment by bashonly** [comment]
> If a user is a guest in a live then it will not provide anything since it only works on that user's specific webcast room. The previous endpoint (sometimes) provides the m3u8 file of the owner of the room where the user is a guest in.

So keeping the old endpoint as a fallback will solve this problem sometimes? Is there anything else we can do about this?

> How often is the _extract_m3u8_formats endpoint called? ... I don't know if it is necessary to call the _extract_m3u8_formats, because I didn't see that it adds any more information

While you're right that the m3u8 is just a media playlist (not master), I've had both API endpoints serve me dead m3u8 URLs, so calling the `_extract_m3u8_formats` method works as a check if the links are still alive. It is called once per m3u8 URL during extraction. ffmpeg is going to be requesting the playlist over and over again during download, so I don't think this is anything to be concerned about

**Comment by bashonly** [comment]
@mokrueger I think I addressed all your suggestions/concerns

**Comment by mokrueger** [comment]
@bashonly Perfect! Thanks for your effort. I think it looks good to me as well. Though, do you know what the proper "protocol" for the flv streams is? I don't know if it's technically required.

**Comment by bashonly** [comment]
> do you know what the proper "protocol" for the flv streams is? I don't know if it's technically required.

http-flv downloads with the native http downloader so it would be `http` or `https` but it doesn't need to be specified

**Comment by mokrueger** [comment]
> http-flv downloads with the native http downloader so it would be `http` or `https` but it doesn't need to be specified

Okay! Thanks for the clarification
